pkg/metadata/pod_hosts.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/prometheus/common/model"
+)
+
+var (
+	initLocalHostname sync.Once
+	hostname          string
+)
+
+// PodHosts provide host ip in default
+// and will provide pod_ip and pod if pid is a pod.
+func PodHosts() Provider {
+	return &StatelessProvider{"hosts", func(pid int) (model.LabelSet, error) {
+		initLocalHostname.Do(func() {
+			local, err := getExternalIPAndHost(1)
+			if err != nil {
+				// this should not happen
+				panic(err)
+			}
+			hostname = local.hostname
+		})
+		e, err := getExternalIPAndHost(pid)
+		if err != nil {
+			return nil, err
+		}
+		if hostname != e.hostname {
+			// pod
+			return model.LabelSet{
+				"pod_ip": model.LabelValue(e.ip),
+				"pod":    model.LabelValue(e.hostname),
+			}, nil
+		}
+		return nil, nil
+	}}
+}
+
+type hostEntry struct {
+	ip       string
+	hostname string
+}
+
+func getExternalIPAndHost(pid int) (*hostEntry, error) {
+	data, err := os.ReadFile(fmt.Sprintf("/proc/%d/root/etc/hosts", pid))
+	if err != nil {
+		return nil, err
+	}
+	hostEntrys, err := parseHosts(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	if len(hostEntrys) == 0 {
+		return nil, fmt.Errorf("read hosts file result nil, raw:%s", string(data))
+	}
+	return &hostEntrys[len(hostEntrys)-1], nil
```

Comment:
Why do we return the pointer here? What's the advantage?

For consistency, either we should stick to the pointers even in the slice or the values. I'd vote for the values since we don't change them at all.

Reply:
Because if return pointer,
Then the return statement could write like `return nil, err ` directly.

I will fix it.